src/librustc_mir/transform/mod.rs

```diff
@@ -267,54 +269,69 @@ fn run_optimization_passes<'tcx>(
     def_id: DefId,
     promoted: Option<Promoted>,
 ) {
+    let post_borrowck_cleanup: &[&dyn MirPass<'tcx>] = &[
+        // Remove all things only needed by analysis
+        &no_landing_pads::NoLandingPads::new(tcx),
+        &simplify_branches::SimplifyBranches::new("initial"),
+        &remove_noop_landing_pads::RemoveNoopLandingPads,
+        &cleanup_post_borrowck::CleanupNonCodegenStatements,
+        &simplify::SimplifyCfg::new("early-opt"),
+        // These next passes must be executed together
+        &add_call_guards::CriticalCallEdges,
+        &elaborate_drops::ElaborateDrops,
+        &no_landing_pads::NoLandingPads::new(tcx),
+        // AddMovesForPackedDrops needs to run after drop
+        // elaboration.
+        &add_moves_for_packed_drops::AddMovesForPackedDrops,
+        // `AddRetag` needs to run after `ElaborateDrops`. Otherwise it should run fairly late,
+        // but before optimizations begin.
+        &add_retag::AddRetag,
+        &simplify::SimplifyCfg::new("elaborate-drops"),
+        // No lifetime analysis based on borrowing can be done from here on out.
+    ];
+
+    let optimizations: &[&dyn MirPass<'tcx>] = &[
+        // Optimizations begin.
+        &unreachable_prop::UnreachablePropagation,
+        &uninhabited_enum_branching::UninhabitedEnumBranching,
+        &simplify::SimplifyCfg::new("after-uninhabited-enum-branching"),
+        &inline::Inline,
+        // Lowering generator control-flow and variables
+        // has to happen before we do anything else to them.
+        &generator::StateTransform,
```

Comment:
This does not look like an optimization, either.

Reply:
I agree, this should be the last thing in the post-borrowck pipeline, it may be run even earlier, not sure, but that requires some experimentation first

Reply:
Moving it to the end of post-borrowck works fine.

Reply:
I think this is deliberately run after some optimizations though as doing the same optimizations post-state-machine-transformation is much harder. Cc @rust-lang/wg-async-foundations

Reply:
The only optimizations before this are

* unreachable_prop
* uninhabited_enum_branching
* SimplifyCfg
* inline

Unreachable propagation works just fine on a desugared state machine.
Uninhabited enum branching would also be no different afaict
There'll still be a `SimplifyCfg` right before it if we move it to then end of post-borrowck
Inlining won't have any effect on  state transform, as inlining always inlines the already optimized MIR, which won't have any state left that could affect the optimization